src/evm/protocol/uniswap_v4/state.rs

```diff
@@ -279,6 +279,79 @@ impl UniswapV4State {
             sqrt_price_next
         }
     }
+
+    fn find_limits_experimentally(
+        &self,
+        token_in: Bytes,
+        token_out: Bytes,
+    ) -> Result<(BigUint, BigUint), SimulationError> {
+        // Create dummy token objects with proper addresses. This is fine since `get_amount_out`
+        // only uses the token addresses.
+        let token_in_obj =
+            Token::new(&token_in, "TOKEN_IN", 18, 0, &[Some(10_000)], Default::default(), 100);
+        let token_out_obj =
+            Token::new(&token_out, "TOKEN_OUT", 18, 0, &[Some(10_000)], Default::default(), 100);
+
+        // Find limit for token_in -> token_out
+        let max_amount_in = self.find_max_amount(&token_in_obj, &token_out_obj)?;
+        // Find limit for token_out -> token_in
+        let max_amount_out = self.find_max_amount(&token_out_obj, &token_in_obj)?;
+
+        Ok((max_amount_in, max_amount_out))
+    }
+
+    /// Finds max amount by performing exponential search.
+    fn find_max_amount(
+        &self,
+        token_in: &Token,
+        token_out: &Token,
+    ) -> Result<BigUint, SimulationError> {
+        let mut low = BigUint::from(1u64);
+
+        // The max you can swap on a USV4 is I256::MAX is 5.7e76, since input amount is I256.
+        // So start with something much smaller to search for a reasonable upper bound.
+        let mut high = BigUint::from(10u64).pow(18); // 1 ether in wei
+        let mut last_successful = BigUint::from(1u64);
+
+        // First, find an upper bound where the swap fails using exponential search.
+        while self
+            .get_amount_out(high.clone(), token_in, token_out)
```

Comment:
Out of the scope of this PR, but the `amount_in` in `get_amount_out` could be passed as a reference right? That would remove a lot of biguint clonings. I'm not very familiar with the biguint type, but it seems to be pretty expensive to clone if the number is high enough.

I would also move the initialization of constant biguints outside of the loops. Even if they are small numbers, they are still allocating/deallocating vec's in each iteration.

Reply:
but thank you for these comments 🙏🏼 and please keep them coming!